xarray/core/dataset.py

```diff
@@ -1575,6 +1575,82 @@ def swap_dims(self, dims_dict, inplace=False):
         return self._replace_vars_and_dims(variables, coord_names,
                                            inplace=inplace)
 
+    def expand_dims(self, dim, axis=None):
+        """Return a new object with an additional axis (or axes) inserted at the
+        corresponding position in the array shape.
+
+        If dim is already a scalar coordinate, it will be promoted to a 1D
+        coordinate consisting of a single value.
+
+        Parameters
+        ----------
+        dim : str or sequence of str.
+            Dimensions to include on the new variable.
+            dimensions are inserted with length 1.
+        axis : integer, list (or tuple) of integers, or None
+            Axis position(s) where new axis is to be inserted (position(s) on
+            the result array). If a list (or tuple) of integers is passed,
+            multiple axes are inserted. In this case, dim arguments should be
+            the same length list or None. If None is passed, all the axes will
+            be inserted to the start of the result array.
+
+        Returns
+        -------
+        expanded : same type as caller
+            This object, but with an additional dimension(s).
+        """
+        # Make sure user does not do `expand_dims(0)`
+        if isinstance(dim, int):
+            raise ValueError('dim should be str or sequence of strs or dict')
+
+        # Make axis and dims a pair of lists
+        if isinstance(dim, basestring):
+            dim = [dim]
+        if axis is not None and not isinstance(axis, (list, tuple)):
+            axis = [axis]
+
+        # infer None arguments.
+        if axis is None:
+            axis = list(range(len(dim)))
+
+        # Error checking
+        # Make sure the length of axis and dims are identical
+        if len(dim) != len(axis):
+            raise ValueError('lengths of dim and axis should be identical.')
+        # make sure any of dim does not already exist
+        for d in dim:
+            if d in self.dims:
+                raise ValueError('{dim} already exists.'.format(dim=d))
+
+        # make sure there is no duplicates in axis or dims
+        if len(axis) != len(set(axis)):
+            raise ValueError('axis should not contain same values.')
+        if len(dim) != len(set(dim)):
+            raise ValueError('dims should not contain same values.')
+
+        # sort dim and axis, so that axis is in order.
+        axis, dim = zip(*sorted(zip(axis, dim)))
+
+        variables = OrderedDict()
+        for k, v in iteritems(self._variables):
+            if k in self._coord_names:
+                if k not in dim: # Do not change coordinates
```

Comment:
Switch the order of these conditions to avoid checking for a negative: it results in clearer code.

Also, [per PEP8](https://www.python.org/dev/peps/pep-0008/#inline-comments) include two spaces before any inline comments.

And per xarray terminology, call these "non-dimension coordinates": http://xarray.pydata.org/en/stable/data-structures.html#coordinates